File: src/main/java/com/mario6/paper/common/ArchiveUtilsExt.java
package com.mario6.paper.common;

import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
import org.apache.commons.compress.archivers.zip.ZipArchiveInputStream;
import org.apache.commons.compress.utils.IOUtils;
import org.apache.commons.lang3.StringUtils;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

/**
 * 压缩相关工具类
 */
public class ArchiveUtilsExt {
    public static final int BUFFER_SIZE = 1024;

    /**
     * 解压 zip 文件
     *
     * @param zipFile zip 压缩文件
     * @param destDir zip 压缩文件解压后保存的目录
     * @return 返回 zip 压缩文件里的文件名的 list
     * @throws Exception
     */
    public static List<String> unZip(File zipFile, String destDir) {
        // 如果 destDir 为 null, 空字符串, 或者全是空格, 则解压到压缩文件所在目录
        if (StringUtils.isBlank(destDir)) {
            destDir = zipFile.getParent();
        }
        destDir = destDir.endsWith(File.separator) ? destDir : destDir + File.separator;
        File destDirFile = new File(destDir);
        if (!destDirFile.exists()) {
            destDirFile.mkdirs();
        }

        String zipPrefixPath = getZipPrefixPath(zipFile);

        ZipArchiveInputStream is = null;
        List<String> fileNames = new ArrayList<String>();

        try {
            is = new ZipArchiveInputStream(new BufferedInputStream(new FileInputStream(zipFile), BUFFER_SIZE));
            ZipArchiveEntry entry = null;

            while ((entry = is.getNextZipEntry()) != null) {
                fileNames.add(entry.getName());

                String contentFileName = entry.getName();
                if (zipPrefixPath != null) {
                    contentFileName = contentFileName.substring(zipPrefixPath.length());
                }

                if (entry.isDirectory()) {
                    File directory = new File(destDir, contentFileName);
                    directory.mkdirs();
                } else {
                    File targetFile = new File(destDir, contentFileName);
                    File targetDir = targetFile.getParentFile();
                    targetDir.mkdirs();
                    OutputStream os = null;
                    try {
                        os = new BufferedOutputStream(new FileOutputStream(targetFile), BUFFER_SIZE);
                        IOUtils.copy(is, os);
                    } finally {
                        IOUtils.closeQuietly(os);
                    }
                }
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        } finally {
            IOUtils.closeQuietly(is);
        }

        return fileNames;
    }

    /**
     * 解压 zip 文件
     *
     * @param zipfile zip 压缩文件的路径
     * @param destDir zip 压缩文件解压后保存的目录
     * @return 返回 zip 压缩文件里的文件名的 list
     * @throws Exception
     */
    public static List<String> unZip(String zipfile, String destDir) {
        File zipFile = new File(zipfile);
        return unZip(zipFile, destDir);
    }

    /**
     * 获取统一前缀
     *
     * @param zipFile
     * @return
     */
    private static String getZipPrefixPath(File zipFile) {
        List<String> paths = getZipContentFilePaths(zipFile);
        String prefix = null;
        for (int i = 0; i < paths.size(); i++) {
            String path = paths.get(i);
            boolean hasPrefix = true;
            if (path.startsWith("__MACOSX")) continue;
            for (int j = i + 1; j < paths.size(); j++) {
                String path2 = paths.get(j);
                if (path2.startsWith("__MACOSX")) continue;
                if (!path2.startsWith(path)) {
                    hasPrefix = false;
                    break;
                }
            }
            if (hasPrefix) {
                prefix = path;
                break;
            }
        }
        return prefix;
    }

    /**
     * 获取指定zip压缩内容文件路径
     *
     * @param zipFile
     * @return
     */
    private static List<String> getZipContentFilePaths(File zipFile) {
        List<String> fileNames = new ArrayList<>();

        ZipArchiveInputStream is = null;
        try {
            is = new ZipArchiveInputStream(new BufferedInputStream(new FileInputStream(zipFile)));
            ZipArchiveEntry entry = null;
            while ((entry = is.getNextZipEntry()) != null) {
                fileNames.add(entry.getName());
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        } finally {
            IOUtils.closeQuietly(is);
        }
        return fileNames;
    }

}

File: src/main/java/com/github/lkqm/paper/util/ArchiveUtilsExt.kt
package com.github.lkqm.paper.util

import org.apache.commons.compress.archivers.zip.ZipArchiveEntry
import org.apache.commons.compress.archivers.zip.ZipArchiveInputStream
import org.apache.commons.compress.utils.IOUtils
import org.apache.commons.lang3.StringUtils
import java.io.*
import java.util.*

/**
 * 压缩相关工具类
 */
object ArchiveUtilsExt {
    const val BUFFER_SIZE = 1024

    /**
     * 解压 zip 文件
     *
     * @param zipFile zip 压缩文件
     * @param destDir zip 压缩文件解压后保存的目录
     * @return 返回 zip 压缩文件里的文件名的 list
     * @throws Exception
     */
    fun unZip(zipFile: File, destDir: String): List<String> {
        // 如果 destDir 为 null, 空字符串, 或者全是空格, 则解压到压缩文件所在目录
        var destDir = destDir
        if (StringUtils.isBlank(destDir)) {
            destDir = zipFile.parent
        }
        destDir = if (destDir.endsWith(File.separator)) destDir else destDir + File.separator
        val destDirFile = File(destDir)
        if (!destDirFile.exists()) {
            destDirFile.mkdirs()
        }
        val zipPrefixPath = getZipPrefixPath(zipFile)
        var `is`: ZipArchiveInputStream? = null
        val fileNames: MutableList<String> = ArrayList()
        try {
            `is` = ZipArchiveInputStream(BufferedInputStream(FileInputStream(zipFile), BUFFER_SIZE))
            var entry: ZipArchiveEntry? = null
            while (`is`.nextZipEntry.also { entry = it } != null) {
                fileNames.add(entry!!.name)
                var contentFileName = entry!!.name
                if (zipPrefixPath != null && contentFileName.startsWith(zipPrefixPath)) {
                    contentFileName = contentFileName.substring(zipPrefixPath.length)
                }
                if (entry!!.isDirectory) {
                    val directory = File(destDir, contentFileName)
                    directory.mkdirs()
                } else {
                    val targetFile = File(destDir, contentFileName)
                    val targetDir = targetFile.parentFile
                    targetDir.mkdirs()
                    var os: OutputStream? = null
                    try {
                        os = BufferedOutputStream(FileOutputStream(targetFile), BUFFER_SIZE)
                        IOUtils.copy(`is`, os)
                    } finally {
                        IOUtils.closeQuietly(os)
                    }
                }
            }
        } catch (e: Exception) {
            throw RuntimeException(e)
        } finally {
            IOUtils.closeQuietly(`is`)
        }
        return fileNames
    }

    /**
     * 解压 zip 文件
     *
     * @param zipfile zip 压缩文件的路径
     * @param destDir zip 压缩文件解压后保存的目录
     * @return 返回 zip 压缩文件里的文件名的 list
     * @throws Exception
     */
    fun unZip(zipfile: String?, destDir: String): List<String> {
        val zipFile = File(zipfile)
        return unZip(zipFile, destDir)
    }

    /**
     * 获取统一前缀
     *
     * @param zipFile
     * @return
     */
    private fun getZipPrefixPath(zipFile: File): String? {
        val paths = getZipContentFilePaths(zipFile)
        var prefix: String? = null
        for (i in paths.indices) {
            val path = paths[i]
            var hasPrefix = false
            if (path.startsWith("__MACOSX") || path.startsWith(".DS_Store")) continue
            for (j in i + 1 until paths.size) {
                val path2 = paths[j]
                if (path2.startsWith("__MACOSX") || path2.startsWith(".DS_Store")) continue
                if (!path2.startsWith(path)) {
                    break
                }
                if (j == paths.size - 1) {
                    hasPrefix = true
                }
            }
            if (hasPrefix) {
                prefix = path
                break
            }
        }
        return prefix
    }

    /**
     * 获取指定zip压缩内容文件路径
     *
     * @param zipFile
     * @return
     */
    private fun getZipContentFilePaths(zipFile: File): List<String> {
        val fileNames: MutableList<String> = ArrayList()
        var `is`: ZipArchiveInputStream? = null
        try {
            `is` = ZipArchiveInputStream(BufferedInputStream(FileInputStream(zipFile)))
            var entry: ZipArchiveEntry? = null
            while (`is`.nextZipEntry.also { entry = it } != null) {
                fileNames.add(entry!!.name)
            }
        } catch (e: Exception) {
            throw RuntimeException(e)
        } finally {
            IOUtils.closeQuietly(`is`)
        }
        return fileNames
    }
}